File: posts/sorting/benchmark.py
#!/usr/bin/env python3
"""
Benchmark script for sorting vs set creation performance comparison.
Updates benchmark_data.json with Python results, preserving JavaScript results.
"""

import json
import time
import random
import gc
from pathlib import Path
import numpy as np

disable_gc = False

def gen_rand_ints(n):
    """Generate n random integers."""
    return [random.random() for _ in range(n)]


def sort_array(arr):
    """Sort array in-place."""
    arr.sort()

def np_sort(arr):
    """Sort array in-place using numpy."""
    np.sort(arr)

def make_set(arr):
    """Create a set from array."""
    set(arr)

def np_set(arr):
    """Create a set from array using numpy."""
    np.unique(arr)

np.sort(np.array([1,2,3])) # warmup

def bench(fn):
    """Benchmark a function and return execution time in milliseconds."""
    start = time.perf_counter()
    fn()
    end = time.perf_counter()
    return (end - start) * 1000  # Convert to milliseconds


def run_benchmarks(use_np=False):
    """Run benchmarks and return results."""
    benchmark_results = []
    num_runs = 10
    
    print("Running Python benchmarks...")
    print(f"Averaging over {num_runs} runs per data point")
    
    # warmup
    sort_array(gen_rand_ints(100000))
    sort_array(np.array(gen_rand_ints(100000)))

    for pow in range(1, 24):
        n = 2 ** pow
        print(f"Testing n = {n:,} (2^{pow})")
        
        total_set_time = 0
        total_sort_time = 0
        
        # Run multiple iterations and average
        for run in range(num_runs):
            # Test Set creation
            arr1 = gen_rand_ints(n)
            arr2 = gen_rand_ints(n)

            if use_np:
                arr1 = np.array(arr1)
                arr2 = np.array(arr2)
                sort_time = bench(lambda: np_sort(arr2))
                set_time = bench(lambda: np_set(arr1))
                total_sort_time += sort_time
            else:
                sort_time = bench(lambda: sort_array(arr2))
                set_time = bench(lambda: make_set(arr1))
                total_sort_time += sort_time
                total_set_time += set_time

        # Calculate averages
        if use_np:
            avg_set_time = None
            avg_sort_time = total_sort_time / num_runs
        else:
            avg_set_time = total_set_time / num_runs
            avg_sort_time = total_sort_time / num_runs

        result = {
            "power": pow,
            "n": n,
            "setTime": avg_set_time,
            "sortTime": avg_sort_time,
            "ratio": avg_sort_time / avg_set_time if avg_set_time and avg_set_time > 0 else 0
        }
        
        benchmark_results.append(result)
        if avg_set_time:
            print(f"  Set: {avg_set_time:.2f}ms, Sort: {avg_sort_time:.2f}ms, Ratio: {result['ratio']:.2f}")
        else:
            print(f"  Sort: {avg_sort_time:.2f}ms")
    
    return benchmark_results


def main():
    """Main function to run benchmarks and update results."""
    # Disable garbage collection for more accurate benchmarking
    if disable_gc:
        gc.disable()
    
    try:
        # Set random seed for reproducible results
        random.seed(42)
        
        # Read existing data or create new structure
        output_file = Path(__file__).parent / "benchmark_data.json"
        data = {}
        
        try:
            if output_file.exists():
                with open(output_file, 'r') as f:
                    existing_data = json.load(f)
                    # Handle both old format (list) and new format (dict with language sections)
                    if isinstance(existing_data, list):
                        # Old format - migrate to new format
                        data = {"python": existing_data}
                    else:
                        # New format - preserve existing data
                        data = existing_data
        except (FileNotFoundError, json.JSONDecodeError):
            print("Creating new benchmark data file or reading existing one")
            data = {}
        
        # Run benchmarks for both regular Python and NumPy
        for use_np in [False, True]:
            language_key = "numpy" if use_np else "python"
            print(f"\n{'='*50}")
            print(f"Running {language_key.upper()} benchmarks...")
            print(f"{'='*50}")
            
            results = run_benchmarks(use_np=use_np)
            data[language_key] = results
            
            print(f"Generated {len(results)} data points for {language_key}")
        
        # Save to JSON file
        with open(output_file, 'w') as f:
            json.dump(data, f, indent=2)
        
        print(f"\nBenchmark results saved to {output_file}")
        print("Results saved for both Python and NumPy")
    
    finally:
        # Re-enable garbage collection
        if disable_gc:
            gc.enable()
            print("Garbage collection re-enabled")


if __name__ == "__main__":
    main() 

File: posts/sorting/benchmark.js
#!/usr/bin/env node
/**
 * Benchmark script for sorting vs set creation performance comparison.
 * Updates benchmark_data.json with JavaScript results, preserving Python results.
 */

const fs = require('fs');
const path = require('path');

function genRandInts(n) {
    /**Generate n random integers.*/
    const arr = [];
    for (let i = 0; i < n; i++) {
        arr.push(Math.random());
    }
    return arr;
}

function sortArray(arr) {
    /**Sort array in-place.*/
    arr.sort((a, b) => a - b);
}

function makeSet(arr) {
    /**Create a Set from array.*/
    new Set(arr);
}

function bench(fn) {
    /**Benchmark a function and return execution time in milliseconds.*/
    const start = performance.now();
    fn();
    const end = performance.now();
    return end - start; // Already in milliseconds
}

function runBenchmarks() {
    /**Run benchmarks and return results.*/
    const benchmarkResults = [];
    const numRuns = 10;
    
    console.log("Running JavaScript benchmarks...");
    console.log(`Averaging over ${numRuns} runs per data point`);
    
    // warmup
    sortArray(genRandInts(100000));

    for (let pow = 1; pow < 24; pow++) {
        const n = 2 ** pow;
        console.log(`Testing n = ${n.toLocaleString()} (2^${pow})`);
        
        let totalSetTime = 0;
        let totalSortTime = 0;
        
        // Run multiple iterations and average
        for (let run = 0; run < numRuns; run++) {
            // Test Set creation
            const arr1 = genRandInts(n);
            const setTime = bench(() => makeSet(arr1));
            totalSetTime += setTime;
            
            // Test sorting  
            const arr2 = genRandInts(n);
            const sortTime = bench(() => sortArray(arr2));
            totalSortTime += sortTime;
        }
        
        // Calculate averages
        const avgSetTime = totalSetTime / numRuns;
        const avgSortTime = totalSortTime / numRuns;
        
        const result = {
            power: pow,
            n: n,
            setTime: avgSetTime,
            sortTime: avgSortTime,
            ratio: avgSetTime > 0 ? avgSortTime / avgSetTime : 0
        };
        
        benchmarkResults.push(result);
        console.log(`  Set: ${avgSetTime.toFixed(2)}ms, Sort: ${avgSortTime.toFixed(2)}ms, Ratio: ${result.ratio.toFixed(2)}`);
    }
    
    return benchmarkResults;
}

function main() {    
    // Run benchmarks
    const results = runBenchmarks();
    
    // Read existing data or create new structure
    const outputFile = path.join(__dirname, "benchmark_data.json");
    let data = {};
    
    try {
        if (fs.existsSync(outputFile)) {
            const existingData = fs.readFileSync(outputFile, 'utf8');
            data = JSON.parse(existingData);
        }
    } catch (error) {
        console.log("Creating new benchmark data file or reading existing one");
        data = {};
    }
    
    // Update only the JavaScript results
    data.javascript = results;
    
    // Save back to JSON file
    fs.writeFileSync(outputFile, JSON.stringify(data, null, 2));
    
    console.log(`\nBenchmark results saved to ${outputFile}`);
    console.log(`Generated ${results.length} data points for JavaScript`);
}

if (require.main === module) {
    main();
}
